main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,255 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+              case BinaryOperator.Divide => Expression.Float32(lit1 / lit2)
+              case BinaryOperator.Exponentiate => Expression.Float32(scala.math.pow(lit1, lit2).toFloat)
+              case BinaryOperator.Less => toExp(lit1 < lit2)
+              case BinaryOperator.LessEqual => toExp(lit1 <= lit2)
+              case BinaryOperator.Greater => toExp(lit1 > lit2)
+              case BinaryOperator.GreaterEqual => toExp(lit1 >= lit2)
+              case _ => Expression.Binary(op, e1, e2, tpe, loc)
+            }
+          // Optimizations for Float64.
+          case (_, Expression.Float64(lit1), Expression.Float64(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case BinaryOperator.Plus => Expression.Float64(lit1 + lit2)
+              case BinaryOperator.Minus => Expression.Float64(lit1 - lit2)
+              case BinaryOperator.Times => Expression.Float64(lit1 * lit2)
+              case BinaryOperator.Divide => Expression.Float64(lit1 / lit2)
+              case BinaryOperator.Exponentiate => Expression.Float64(scala.math.pow(lit1, lit2))
+              case BinaryOperator.Less => toExp(lit1 < lit2)
+              case BinaryOperator.LessEqual => toExp(lit1 <= lit2)
+              case BinaryOperator.Greater => toExp(lit1 > lit2)
+              case BinaryOperator.GreaterEqual => toExp(lit1 >= lit2)
+              case _ => Expression.Binary(op, e1, e2, tpe, loc)
+            }
+          // Optimizations for Int8.
+          case (_, Expression.Int8(lit1), Expression.Int8(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case BinaryOperator.Plus => Expression.Int8((lit1 + lit2).toByte)
+              case BinaryOperator.Minus => Expression.Int8((lit1 - lit2).toByte)
+              case BinaryOperator.Times => Expression.Int8((lit1 * lit2).toByte)
+              case BinaryOperator.Divide => Expression.Int8((lit1 / lit2).toByte)
+              case BinaryOperator.Exponentiate => Expression.Int8(scala.math.pow(lit1, lit2).toByte)
+              case BinaryOperator.Less => toExp(lit1 < lit2)
+              case BinaryOperator.LessEqual => toExp(lit1 <= lit2)
+              case BinaryOperator.Greater => toExp(lit1 > lit2)
+              case BinaryOperator.GreaterEqual => toExp(lit1 >= lit2)
+              case _ => Expression.Binary(op, e1, e2, tpe, loc)
+            }
+          // Optimizations for Int16.
+          case (_, Expression.Int16(lit1), Expression.Int16(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case BinaryOperator.Plus => Expression.Int16((lit1 + lit2).toShort)
```

Comment:
This is correct for Flix. The `toShort` (and `toByte` for `Int8`) is crucial. Without it, the result is sign-extended to an `int`, which might not fit in an `Int8`.